notifications/v1/client.go

```diff
@@ -6,90 +6,153 @@ import (
 
 	"cloud.google.com/go/pubsub"
 	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/lib/background"
 	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"go.uber.org/atomic"
 )
 
-// Client is a client for receiving SAMS notifications from a GCP Pub/Sub
-// subscription.
-type Client struct {
-	logger        log.Logger
-	subscription  *pubsub.Subscription
+type subscriber struct {
+	logger       log.Logger
+	handlers     SubscriberHandlers
+	subscription *pubsub.Subscription
+
+	// state indicates the state of workers, must be one of "idle", "started" or
+	// "stopped".
+	state *atomic.String
+	// cancelContext is the function to cancel the context of the receiver that
+	// effectively stops the receiver.
 	cancelContext context.CancelFunc
 }
 
-// NewClient creates a new Client for receiving SAMS notifications with given
-// GCP project ID and Pub/Sub subscription ID.
-func NewClient(logger log.Logger, projectID, subscriptionID string) (*Client, error) {
-	client, err := pubsub.NewClient(context.Background(), projectID)
+type SubscriberOptions struct {
+	// ProjectID is the GCP project ID that the Pub/Sub subscription belongs to. It
+	// is almost always the same GCP project that the Cloud Run service is deployed
+	// to.
+	ProjectID string
+	// SubscriptionID is the GCP Pub/Sub subscription ID to receive SAMS
+	// notifications from.
+	SubscriptionID string
+	// ReceiveSettings is the settings for receiving messages of the subscription. A
+	// zero value means to use the default settings.
+	ReceiveSettings ReceiveSettings
+	// Handlers is the collection of subscription handlers for each type of SAMS
+	// notifications.
+	Handlers SubscriberHandlers
+}
+
+func (opts SubscriberOptions) Validate() error {
+	if opts.ProjectID == "" {
+		return errors.New("ProjectID is required")
```

Comment:
Kinda surprised the linter isn't complaining that the error message starts with a capital letter. (i.e. this error is "good", but a situation where linters usually flag it as a false positive... so usually I find that I need to write something lame like "**property** X is required".)

But maybe linters are smart enough now?

Reply:
Haha yeah, sometimes the linear is smart enough (assuming we did enable the one you're referring to 😂